### DISCUSSION TITLE
Is it possible to pass storage config dynamically ?

### DISCUSSION THREAD

**Discussion by troptropcontent** [shrinerb/shrine#669] [General]
Hello

Happy new year, i hope that you are doing well

I would like to let my users decide in wich storage they want to storage there files (by providing there credentisl)

Is it possible to pass credentials at the uploader lever ? 

To do something like 

`MyUploader.upload(file, {credentials})`

? 

Thank you so much for you help

**Comment by troptropcontent** [comment]
Oups sorry i did not see that there where a discussion !!

I might try something like this 

```ruby
unique_id_key = rand()
class MyUploader < Shrine;end
uploader = MyUploader.new(unique_id_key.to_sym)
# And here i touch the singleton of uploader to add the unique_id_key storage to the storages
```

Do you think that i could work ? 🤔

**Reply by janko** [reply]
I converted your issue into a discussion, as issues are only intended for bug reports 😉 

If your users would create AWS credentials as database records, and if you used Active Support's current attributes, you could do something like this:

```rb
Shrine.plugin :default_storage,
  cache: -> { Current.user ? :"cache_#{Current.user.credential.id}" : :cache }
  store: -> { Current.user ? :"store_#{Current.user.credential.id}" : :store }

Shrine.plugin :dynamic_storage
Shring.storage /cache_(\w+)/ do |match|
  credential = Credential.find(match[1])
  Shrine::Storage::S3.new(...) # initialize storage from credential data
end
Shring.storage /store_(\w+)/ do |match|
  credential = Credential.find(match[1])
  Shrine::Storage::S3.new(...) # initialize storage from credential data
end
```

**Reply by troptropcontent** [reply]
This is perfect !!!

Thank you so much for the prompt response
